torchao/quantization/subclass.py

```diff
@@ -592,3 +604,263 @@ def to_qtensor_components(cls, input_float, groupsize=128, inner_k_tiles=8):
+            memory_format if memory_format is not None else torch.preserve_format
+        )
+        kwargs = {
+            "device": device,
+            "dtype": dtype,
+            "memory_format": memory_format,
+        }
+        return kwargs
+
+    def to(self, *args, **kwargs):
+        kwargs = self._get_to_kwargs(*args, **kwargs)
+        return self.__class__(
+            self.int_data.to(kwargs["device"]),
+            self.scale.to(kwargs["device"]),
+            self.zero_point.to(kwargs["device"]),
+            self.block_size,
+            self.shape,
+            self.quant_min,
+            self.quant_max,
+            self.input_quant_func,
+            **kwargs,
+        )
+
+    def _apply_fn_to_data(self, fn):
+        return self.__class__(
+            fn(self.int_data),
+            fn(self.scale),
+            fn(self.zero_point),
+            self.block_size,
+            self.shape,
+            self.quant_min,
+            self.quant_max,
+            self.input_quant_func,
+            dtype=self.dtype,
+        )
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        # two scenarios where we currently fall back to vanilla mm:
+        # 1 - when tensor is on CPU: we are missing qmm for CPU, but we should have a CPU implementation
+        #     for consistency and to allow people to test
+        # 2 - we're given non-floats - quantizing long to int8 is crazy
+        if (
+            func in [aten.mm.default, aten.addmm.default]
+            and args[0].is_floating_point()
+            and args[0].is_cuda
+        ):
+            if func == aten.addmm.default:
+                assert args[1].shape[-1] == args[2].shape[0], (
+                    f"need mat1 shape: {args[1].shape} final"
+                    f"dim to match mat2 shape: {args[2].shape} first dim "
+                )
+                input_tensor, weight_qtensor, bias = (
+                    args[1],
+                    args[2],
+                    args[0],
+                )
+            else:
+                assert args[0].shape[-1] == args[1].shape[0], (
+                    f"need mat1 shape: {args[0].shape} final dim"
+                    f"to match mat2 shape: {args[1].shape} first dim"
+                )
+                input_tensor, weight_qtensor, bias = (
+                    args[0],
+                    args[1],
+                    None if len(args) == 2 else args[2],
+                )
+            if weight_qtensor.input_quant_func is not None:
+                input_tensor = weight_qtensor.input_quant_func(input_tensor)
+                input_tensor = input_tensor.dequantize()
+            weight_tensor = weight_qtensor.dequantize()
+            return func(input_tensor, weight_tensor, bias)
+
+        if (func is aten.detach.default or
+            func is aten.clone.default or
+            func is aten._to_copy.default):
+            return return_and_correct_aliasing(
```

Comment:
These are the kind of things that make me wonder if AffineQuantizedTensor via tensor subclass is the right approach. 

Are we trying to unify all kind of quant techniques behind this class? If so, I am not sure if thats a good idea

Reply:
So, I think we should still have a legokit of ops for quantization (e.g. (mixed) integer matmul, affine quant/dequant primitives, etc.), but at least historically we've used dtypes for lower precision data types (e.g. bfloat16). AQT is a way to provide a unified interface for dtypes based on affine quantization. People don't *have* to use it of course, but it seems convenient for like trying a new op (for example see https://github.com/pytorch/ao/blob/ad12663c736c3ca5960fbb3a9b8f7df2d4d65b3a/tutorials/add_an_op.py#L24 ) or just printing the content of a quantized tensor.

Reply:
I see. This is similar to what @jerryzh168 was suggesting about allowing users to override dispatch. 

I think my main sticking point was this: the dynamic quantization seems like an aberration to the relatively ok patterns of "registering" in op for the tensor subclass as you have shown in the gelu example. Because in dynamic quant, at least for linear, we are saying I want to quantized input activation, but if i am doing weight only quant then i dont want to quantize input activation. And we are trying to encode this information in AQT via layout and other things.

Reply:
@kimishpatel - Yes, I agree. We should separate "injection" from the dtype of affine integer quantization. 

Dynamic quantization is the process of recalculating the range and zero point of an affine integer quantized tensor on every call. It's static if you just use precalculated ranges and zero points. But morally the dtype is still an affine integer quantized tensor. It's just a particular way of representing real numbers in memory. 

"injection" can be done in multiple ways. We could use tensor subclasses and reparameterize to e.g. replace F.linear with a couple ops. Or you can swap modules. Some users might even just modify the model itself and change the dtype of layers (like model.half()). Then we could also use FX passes and such to update the IR and inject this dtype. But that's separate.